### DISCUSSION TITLE
Point queries

### DISCUSSION THREAD

**Discussion by fluffysquirrels** [bevyengine/bevy#3169] [Q&A] [answered]
I'm trying to do a point query in one of my systems, by which I mean looking up
a component belonging to an entity that I know the entity id of.

Here's my code so far:

```rust
pub type Hp = i32;
pub struct HealthComponent {
    pub current: Hp,
    pub max: Hp,
}
pub struct Player;
pub struct DamageEvent {
    pub target: Entity,
    pub hp: Hp,
}

fn setup(
    mut commands: Commands,
) {
    commands
        .spawn()
        .insert(Player)
        .insert(HealthComponent {
            current: 10,
            max: 10,
        }

}

fn apply_damage(
    mut commands: Commands,
    dmg_reader: mut EventReader<DamageEvent>,
) {
    for damage in dmg_reader.iter() {
        let entity = commands.entity(damage.target);
        let mut health: HealthComponent = todo!();
        health.current -= damage.hp;
    }
}
```

The bit I'm uncertain of is what to replace the `todo!()` with.

If I had access to a `World` I could just do
`world.entity(entity).get::<HealthComponent>()`, but I don't know how
to get access to a `World`. I suppose I could stash the `World` while
using my `AppBuilder` but that doesn't seem very ergonomic.

`Commands` lets me modify the entity but not query it. Is there a
reason for this? Perhaps ad-hoc access to components prevents
lock-free parallelisation?

I could inject a `Query` into `apply_damage` to loop over every entity
with a `HealthComponent` component, but that sounds like it would
require `O(^2)` iterations where `n` is the number of entities that
have a `HealthComponent` component.

Perhaps there is a more idiomatic way to represent `HealthComponent` or `DamageEvent`?

Any help is appreciated.

**Comment by TheRawMeatball** [comment, marked as answer]
You can use query.get or query.get_component :)

**Reply by fluffysquirrels** [reply]
Ooops! Thanks!

**Reply by tatref** [reply]
Here is a complete example:

```rust
fn apply_damage(
    mut commands: Commands,
    dmg_reader: mut EventReader<DamageEvent>,
    mut q: Query<&mut HealthComponent>
) {
    for damage in dmg_reader.iter() {
        let entity = commands.entity(damage.target);
        let mut health = q.get_mut(entity).expect("Entity not found or no HealthComponent");
        health.current -= damage.hp;
    }
}
```

You can also use `if let Some(health) = q.get_mut(...` for better error handeling.
